game/audio.cc

```diff
@@ -62,173 +67,191 @@ namespace {
-	bool operator()(float* begin, float* end) {
-		size_t samples = end - begin;
-		m_clock.timeSync(durationOf(m_pos), durationOf(samples)); // Keep the clock synced
-		bool eof = true;
-		Buffer mixbuf(samples);
-		for (auto& kv: tracks) {
-			Track& t = *kv.second;
-// FIXME: Include this code bit once there is a sane pitch shifting algorithm
-#if 0
-//			if (it->first == "guitar") std::cout << t.pitchFactor << std::endl;
-			if (t.pitchFactor != 0) { // Pitch shift
-				Buffer tempbuf(end - begin);
-				// Get audio to temp buffer
-				if (t.mpeg.audioQueue(&*tempbuf.begin(), &*tempbuf.end(), m_pos, t.fadeLevel)) eof = false;
-				// Do the magic
-				PitchShift(&*tempbuf.begin(), &*tempbuf.end(), t.pitchFactor);
-				// Mix with other tracks
-				Buffer::iterator m = mixbuf.begin();
-				Buffer::iterator b = tempbuf.begin();
-				while (b != tempbuf.end())
-					*m++ += (*b++);
-			// Otherwise just get the audio and mix it straight away
-			} else
-#endif
-			if (t.mpeg.audioQueue(&*mixbuf.begin(), &*mixbuf.end(), m_pos, t.fadeLevel)) eof = false;
-		}
-		m_pos += samples;
-		for (size_t i = 0, iend = mixbuf.size(); i != iend; ++i) {
-			if (i % 2 == 0) {
-				fadeLevel += fadeRate;
-				if (fadeLevel <= 0.0) return false;
-				if (fadeLevel > 1.0) { fadeLevel = 1.0; fadeRate = 0.0; }
-			}
-			begin[i] += mixbuf[i] * fadeLevel * static_cast<float>(m_preview ? config["audio/preview_volume"].i() : config["audio/music_volume"].i())/100.0;
-		}
-		// suppress center channel vocals
-		if(suppressCenterChannel && !m_preview) {
-			float diffLR;
-			for (size_t i=0; i<mixbuf.size(); i+=2) {
-				diffLR = begin[i] - begin[i+1];
-				begin[i] = diffLR;
-				begin[i+1] = diffLR;
-			}
-		}
-		return !eof;
+	suppressCenterChannel = config["audio/suppress_center_channel"].b();
+	aubio_tempo_set_silence(aubioTempo.get(), -41.0);
+	aubio_tempo_set_threshold(aubioTempo.get(), 1.25);
+}
+
+unsigned Audio::aubio_win_size = 2048;
+unsigned Audio::aubio_hop_size = 1024;
+
+std::unique_ptr<aubio_tempo_t, void(*)(aubio_tempo_t*)> Music::aubioTempo = std::unique_ptr<aubio_tempo_t, void(*)(aubio_tempo_t*)>(new_aubio_tempo("default", Audio::aubio_win_size, Audio::aubio_hop_size, Audio::getSR()),[](aubio_tempo_t* p) {
+	if (p != nullptr) {
+		del_aubio_tempo(p);
 	}
-	void seek(double time) { m_pos = time * srate * 2.0; }
-	/// Get the current position in seconds
-	double pos() const { return m_clock.pos().count(); }
-	double duration() const {
-		double dur = 0.0;
-		for (auto& kv: tracks) dur = std::max(dur, kv.second->mpeg.audioQueue.duration());
-		return dur;
+});
+std::recursive_mutex Music::aubio_mutex;
+
+bool Music::operator()(float* begin, float* end) {
+	size_t samples = end - begin;
+	m_clock.timeSync(durationOf(m_pos), durationOf(samples)); // Keep the clock synced
+	bool eof = true;
+	Buffer mixbuf(samples);
+	for (auto& kv: tracks) {
+		Track& t = *kv.second;
+// #if 0 // FIXME: Include this code bit once there is a sane pitch shifting algorithm
```

Comment:
Define `sane pitch shifting algorithm` and remove this from the code and create an issue. This change can always be reverted by git so there's really no use in keeping commented code blocks in the code.

Reply:
This has been there forever, I just commented-it-out to help us visually on the syntax highlighting. The thing is, we don't have any implementations for pitch shifting, but we do have (stub) functions and structures for it. I'd just leave it there until pitch-shifting has been merged into aubio master and then we can use that.